src/Service/Bulk/BulkUserCreationService.php

```diff
@@ -0,0 +1,324 @@
+use OAT\SimpleRoster\Repository\Criteria\FindLineItemCriteria;
+use OAT\SimpleRoster\DataTransferObject\UserCreationResult;
+use OAT\SimpleRoster\Csv\BulkUserCreationCsvWriter;
+use Psr\Log\LoggerInterface;
+
+class BulkUserCreationService
+{
+    private array $lineItemIds = [];
+    private array $lineItemSlugs = [];
+
+    private int $userGroupBatchCount = 0;
+    private int $groupIndex = 0;
+
+    private string $generatedUsersFilePath;
+
+    private LineItemRepository $lineItemRepository;
+    private AssignmentRepository $assignmentRepository;
+    private Filesystem $filesystem;
+    private GroupIdLoadBalancerFactory $groupIdLoadBalancerFactory;
+    private UserAssignmentIngester $userAssignmentIngester;
+    private BulkUserCreationCsvWriter $csvWriter;
+    private LoggerInterface $logger;
+
+    private const DEFAULT_USERNAME_INCREMENT_VALUE = 0;
+
+    public function __construct(
+        LineItemRepository $lineItemRepository,
+        AssignmentRepository $assignmentRepository,
+        GroupIdLoadBalancerFactory $groupIdLoadBalancerFactory,
+        UserAssignmentIngester $userAssignmentIngester,
+        BulkUserCreationCsvWriter $csvWriter,
+        LoggerInterface $logger,
+        Filesystem $filesystem,
+        string $generatedUsersFilePath
+    ) {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->assignmentRepository = $assignmentRepository;
+        $this->groupIdLoadBalancerFactory = $groupIdLoadBalancerFactory;
+        $this->userAssignmentIngester = $userAssignmentIngester;
+        $this->csvWriter = $csvWriter;
+        $this->logger = $logger;
+        $this->generatedUsersFilePath = $generatedUsersFilePath;
+        $this->filesystem = $filesystem;
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function createUsers(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefixes,
+        int $batchSize,
+        ?string $groupPrefix
+    ): UserCreationResult {
+
+        $notExistLineItemsArray = $userGroupIds = [];
+
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $this->lineItemIds = $lineItemIds;
+            $this->lineItemSlugs = $lineItemSlugs;
+            $criteria = $this->getFindLineItemCriteria();
+            $lineItems = $this->lineItemRepository->findLineItemsByCriteria($criteria)->jsonSerialize();
+            if (empty($lineItems)) {
+                $exceptionMessage = $this->lineItemIds
+                    ? implode(',', $this->lineItemIds) . ' Line item id(s) not exist in the system'
+                    : implode(',', $this->lineItemSlugs) . ' Line item slug(s) not exist in the system';
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $this->generateSlugData($lineItems);
+            $notExistLineItemsArray = $this->lineItemIds
+                ? array_diff($lineItemIds, array_keys($this->lineItemSlugs))
```

Comment:
The array key of $this->lineItemSlugs is the line item DB id. So in the case of $lineItemIds option need to check with array_keys($this->lineItemSlugs) and for $lineItemSlugs need to check with array_values($this->lineItemSlugs)

Reply:
$this->lineItemSlugs array key contains line item DB id. so in the case of $this->lineItemIds it will check with array_keys($this->lineItemSlugs) and in the case of $this->lineItemSlugs will check with array_values($this->lineItemSlugs)